### PR TITLE
SONARPY-143 Support issues with precise location

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java [hunk 1]
```diff
@@ -93,9 +100,48 @@ public void analyse(Project project, SensorContext context) {
...
+      visitors
+        .stream()
+        .filter(v -> v instanceof PythonCheck)
+        .map(v -> (PythonCheck) v)
+        .collect(Collectors.toList()),
+      context);
     save(squidSourceFiles);
   }
 
+  // fixme (Lena): unit test it with new Sensor API
```
Comment (SonarTech) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java [hunk 2]
```diff
@@ -93,9 +100,48 @@ public void analyse(Project project, SensorContext context) {
...
+  private void savePreciseIssues(List<PythonCheck> pythonChecks, SensorContext context) {
```
Comment (SonarTech) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/PythonSquidSensor.java]: MINOR Remove this use of "SensorContext"; it is deprecated. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+      fail("Unexpected issue at line " + line(actual) + ": \"" + actual.primaryLocation().message() + "\"");
+    }
+    if (expected.message() != null) {
+      assertThat(actual.primaryLocation().message()).as("Bad message at line " + expected.line()).isEqualTo(expected.message());
+    }
+    if (expected.effortToFix() != null) {
+      assertThat(actual.cost()).as("Bad effortToFix at line " + expected.line()).isEqualTo(expected.effortToFix());
+    }
+    if (expected.startColumn() != null) {
+      assertThat(((PreciseIssue) actual).primaryLocation().startLineOffset() + 1).as("Bad start column at line " + expected.line()).isEqualTo(expected.startColumn());
```
Comment (pynicolas) [inline python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java]: @vilchik-elena Eclipse tells me that there is an unnecessary cast to PreciseIssue here and on lines 89 and 92.

File: python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      if (equalIndex == -1) {
+        throw new IllegalStateException("Invalid param at line 1: " + param);
+      }
+      String name = param.substring(0, equalIndex);
+      String value = param.substring(equalIndex + 1);
+
+      if ("effortToFix".equalsIgnoreCase(name)) {
+        issue.effortToFix(Integer.valueOf(value));
+
+      } else if ("sc".equalsIgnoreCase(name)) {
```
Comment (pynicolas) [inline python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java]: @vilchik-elena Do we want to support such feature?

Comment (vilchik-elena) [inline python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java]: @pynicolas I think so. E.g. if we want to test issue located at several lines

File: python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java (hunk 3 above)
Comment (vilchik-elena) [inline python-checks/src/test/java/org/sonar/python/checks/utils/PythonCheckVerifier.java]: @pynicolas thanks! fixed d4cd172d261fa9df69e2d42106ebd3e5c05e039b

**Pull request merged**